src/core/broker_config.cpp

```diff
@@ -17,37 +17,207 @@
  * https://github.com/ETCLabs/RDMnetBroker
  *****************************************************************************/
 
+#define NOMINMAX
 #include "broker_config.h"
 
+#include <functional>
 #include <fstream>
+#include <limits>
+#include <map>
+#include "etcpal/uuid.h"
 
-constexpr char* kDefaultSettings = R"(
+// A set of information needed to validate an element in the config file's JSON.
+struct Validator
+{
+  // Pointer to the key
+  const json::json_pointer pointer;
+  // The expected value type
+  const json::value_t type;
+  // A function to validate and store the value in the config's settings structure.
+  std::function<bool(const json&, BrokerConfig&)> validate_and_store;
+  // A function to store the default setting in the settings structure if the value is not present
+  // in the config file.
+  std::function<void(BrokerConfig&)> store_default;
+};
+
+// The CID must be a string representation of a UUID.
+bool ValidateAndStoreCid(const json& val, BrokerConfig& config)
+{
+  std::string str_val = val;
+  return etcpal_string_to_uuid(&config.settings.cid, str_val.c_str(), str_val.size());
+}
+
+// The UID takes the form:
+// "uid": {
+//   "type": < "static" | "dynamic" >,
+//   "manufacturer_id": <number, always present>,
+//   "device_id": <number, present only if type is "static">
+// }
+bool ValidateAndStoreUid(const json& val, BrokerConfig& config)
+{
+  if (val.contains("type") && val["type"].is_string() && val.contains("manufacturer_id") &&
+      val["manufacturer_id"].is_number_integer())
   {
-    "cid": "",
-    "uid_type": "dynamic",
-    "uid": "",
+    std::string type = val["type"];
+    int64_t manufacturer_id = val["manufacturer_id"];
 
-    "dns_sd": {
-      "service_instance_name": "",
-      "manufacturer": "ETC",
-      "model": "RDMnet Broker",
-      "scope": ""
-    },
+    if (manufacturer_id > 0 && manufacturer_id < 0x8000)
+    {
+      if (type == "static" && val.contains("device_id") && val["device_id"].is_number_integer())
+      {
+        int64_t device_id = val["device_id"];
+        if (device_id >= 0 && device_id <= 0xffffffff)
+        {
+          RdmUid static_uid;
+          static_uid.manu = static_cast<uint16_t>(manufacturer_id);
+          static_uid.id = static_cast<uint32_t>(device_id);
+          config.settings.SetStaticUid(static_uid);
+          return true;
+        }
+      }
+      else if (type == "dynamic" && !val.contains("device_id"))
+      {
+        config.settings.SetDynamicUid(static_cast<uint16_t>(manufacturer_id));
+        return true;
+      }
+    }
+  }
+  return false;
+}
 
-    "max_connections": 0,
-    "max_controllers": 0,
-    "max_controller_messages": 500,
-    "max_devices": 0,
-    "max_device_messages": 500,
-    "max_reject_connections": 1000
+// Store a generic string from the "dns_sd": {} object.
+template <size_t TruncationSize>
```

Comment:
This is cool.  I've never overridden a template with a size_t parameter before!

Reply:
I got the idea from `std::array`. But looking at this again, I think I'm just going to make the truncation size a function parameter. I originally thought by making it a template I would be able to avoid having an intermediate lambda in the settings array below, but I still need one, so the value of the template is kind of lost.